### PR TITLE
FOUR-9950 Ability to create an asset from the Modeler (Asset Quick Create)

### PR DESCRIPTION
**Pull request by josechirivella**
## Solution
- Added a link in Screens, Scripts, and Decision Tables to create those from the Modeler itself by opening a tab
- Created a Component to handle this logic and used an enum to validate the input of those
- Added a Broadcast Channel to get/receive data between browser tabs to communicate when an asset was created

## Related Tickets & Packages
- https://processmaker.atlassian.net/browse/FOUR-9949
- https://processmaker.atlassian.net/browse/FOUR-9950
- https://processmaker.atlassian.net/browse/FOUR-9951
- https://processmaker.atlassian.net/browse/FOUR-9952
- https://processmaker.atlassian.net/browse/FOUR-9953

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:deploy
ci:package-decision-engine:FOUR-9953
ci:package-projects:next
..

### CODE REVIEW

File: resources/js/bootstrap.js [hunk 1]
```diff
@@ -42,6 +42,10 @@ window.ProcessmakerComponents = require("./processes/screen-builder/components")
  */
 window.SharedComponents = require("./components/shared");
 
+/**
+ * Exporting Modeler inspector components
+ */
+window.ModelerInspector = require("./processes/modeler/components/inspector");
```
Comment (josechirivella) [inline resources/js/bootstrap.js:48]: I needed to export this component to be used in the package decision tables

File: resources/js/processes/modeler/components/inspector/ModelerAssetQuickCreate.vue [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  </div>
+</template>
+<script>
+import { kebabCase } from "lodash";
+
+const AssetTypes = Object.freeze({
+  SCREEN: "screen",
+  SCRIPT: "script",
+  DECISION_TABLE: "decision table",
+});
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ModelerAssetQuickCreate.vue:19]: Enum

File: resources/js/processes/modeler/components/inspector/ModelerAssetQuickCreate.vue [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+const channel = new BroadcastChannel("assetCreation");
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ModelerAssetQuickCreate.vue:21]: You can read more about it, here https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API

File: resources/js/processes/modeler/components/inspector/ScreenSelect.vue
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScreenSelect.vue:1]: Mostly indentation of this file but I will point where the logic lives

File: resources/js/processes/modeler/components/inspector/ScreenSelect.vue [hunk 4]
```diff
@@ -1,151 +1,175 @@
...
+      <div>{{ error }}</div>
     </div>
+    <small
+      v-if="helper"
+      class="form-text text-muted"
+    >{{ $t(helper) }}</small>
+    <modeler-asset-quick-create
+      label="screen"
+      @asset="processAssetCreation"
+    />
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScreenSelect.vue:43]: added here the component I created with their emitter

File: resources/js/processes/modeler/components/inspector/ScreenSelect.vue [hunk 5]
```diff
@@ -1,151 +1,175 @@
...
+        });
+    },
+    /**
+     * @param {Object} data - The response we get from the emitter
+     * @param {string} data.id - the screen id
+     * @param {string} data.assetType - The Asset type, ex: screen
+     */
+    processAssetCreation({ id, assetType }) {
+      if (assetType === "screen") this.$emit("input", id);
+    },
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScreenSelect.vue:161]: process the emitter from the AssetQuickCreate component and sent an emit to the multiselect to select the value from the screen creaed

File: resources/js/processes/modeler/components/inspector/ScriptSelect.vue
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScriptSelect.vue:1]: Mostly indentation but I will point out where the changes happened

File: resources/js/processes/modeler/components/inspector/ScriptSelect.vue [hunk 6]
```diff
@@ -1,149 +1,174 @@
...
+      <div>{{ error }}</div>
     </div>
+    <small
+      v-if="helper"
+      class="form-text text-muted"
+    >{{ $t(helper) }}</small>
+    <modeler-asset-quick-create
+      label="script"
+      @asset="processAssetCreation"
+    />
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScriptSelect.vue:42]: Newly created Component here

File: resources/js/processes/modeler/components/inspector/ScriptSelect.vue [hunk 7]
```diff
@@ -1,149 +1,174 @@
...
+    /**
+     * @param {Object} data - The response we get from the emitter
+     * @param {string} data.id - the screen id
+     * @param {string} data.assetType - The Asset type, ex: screen
+     */
+    processAssetCreation({ id, assetType }) {
+      if (assetType === "script") this.$emit("input", id);
+    },
+  },
+};
```
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/ScriptSelect.vue:169]: Processing the emitter and sending the created script to multiselect

File: resources/js/processes/screens/components/CreateScreenModal.vue
Comment (josechirivella) [inline resources/js/processes/screens/components/CreateScreenModal.vue:1]: Mostly indentation but I will point out where the changes happened

File: resources/js/processes/modeler/components/inspector/index.js
Comment (josechirivella) [inline resources/js/processes/modeler/components/inspector/index.js:1]: Needed to export this to be used in the package (decision engine)

File: resources/js/processes/screens/components/CreateScreenModal.vue [hunk 8]
```diff
@@ -1,129 +1,164 @@
...
-  };
+      this.disabled = true;
+      ProcessMaker.apiClient
+        .post("screens", this.formData)
+        .then(({ data }) => {
+          ProcessMaker.alert(this.$t("The screen was created."), "success");
+          channel.postMessage({
+            assetType: "screen",
+            id: data.id,
+          });
```
Comment (josechirivella) [inline resources/js/processes/screens/components/CreateScreenModal.vue:152]: post the message via the Broadcast Channel that a new screen was created

File: resources/js/processes/screens/index.js [hunk 9]
```diff
@@ -16,6 +16,12 @@ new Vue({
     screenModal: false,
     screenId: null,
   },
+  mounted() {
+    const searchParams = new URLSearchParams(window.location.search);
+    if (searchParams.size > 0 && searchParams.get("create") === "true") {
+      this.$root.$emit("bv::show::modal", "createScreen", "#createScreenModalButton");
+    }
+  },
```
Comment (josechirivella) [inline resources/js/processes/screens/index.js:24]: Using URL queries to establish that I want to create a new screen and the modal should open automatically

File: resources/js/processes/scripts/components/CreateScriptModal.vue
Comment (josechirivella) [inline resources/js/processes/scripts/components/CreateScriptModal.vue:1]: Mostly indentation

File: resources/js/processes/scripts/components/CreateScriptModal.vue [hunk 10]
```diff
@@ -176,24 +241,28 @@
...
-  };
+        .then(({ data }) => {
+          ProcessMaker.alert(this.$t("The script was created."), "success");
+          (this.$refs.createScriptHooks || []).forEach((hook) => {
+            hook.onsave(data);
+          });
+          channel.postMessage({
+            assetType: "script",
+            id: data.id,
+          });
```
Comment (josechirivella) [inline resources/js/processes/scripts/components/CreateScriptModal.vue:254]: posting the message via the Broadcast Channel that a new script was created

File: resources/js/processes/scripts/index.js [hunk 11]
```diff
@@ -14,6 +14,12 @@ new Vue({
   data: {
     filter: "",
   },
+  mounted() {
+    const searchParams = new URLSearchParams(window.location.search);
+    if (searchParams.size > 0 && searchParams.get("create") === "true") {
+      this.$root.$emit("bv::show::modal", "createScript", "#createScriptModalButton");
+    }
+  },
```
Comment (josechirivella) [inline resources/js/processes/scripts/index.js:22]: Auto-open the new script modal using URL queries

**Review by josechirivella — COMMENTED** [review]
Some guidance comments

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-ab61860519.eng.processmaker.net

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-ab61860519.eng.processmaker.net

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-ab61860519.eng.processmaker.net

**Review by danloa — APPROVED** [review]

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-ab61860519.eng.processmaker.net

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-ab61860519.eng.processmaker.net

**Pull request merged**